File: typescript/secret-handshake/secret-handshake.ts
const dict = [
	{ binary: 1000, value: 'jump', },
	{ binary: 100, value: 'close your eyes', },
	{ binary: 10, value: 'double blink', },
	{ binary: 1, value: 'wink', },
]

const reverseValue = 10000;

export function commands(decimal: number) {
	let binary = decimalToBinary(decimal);
	if (binary <= 0) {
		return [];
	}

	const reverse = binary - reverseValue >= 0;
	if (reverse) {
		binary -= reverseValue;
	}

	const operations: string[] = [];
	let stop: boolean;
	do {
		stop = true;
		dict.forEach(({ binary: n, value }) => {
			if (binary - n >= 0) {
				stop = false;
				binary -= n;
				operations.unshift(value);
			}
		});

	} while (binary > 0 && !stop);

	if (reverse) {
		return operations.reverse();
	}

	return operations;
}

const decimalToBinary = (decimal: number): number => {
	return Number((decimal >>> 0).toString(2));
}

File: go/secret-handshake/secret_handshake.go
package secret

import (
	"slices"
	"strconv"
)

const (
	wink          = "wink"
	doubleBlink   = "double blink"
	closeYourEyes = "close your eyes"
	jump          = "jump"
)

type modifier struct {
	index  int
	action string
	modify func(binary string, c []string) []string
}

func (m *modifier) apply(binary string, c []string) []string {
	if len(binary) < m.index {
		return c
	}

	if binary[len(binary)-m.index] == '1' {
		if m.modify != nil {
			c = m.modify(binary, c)
		} else {
			c = append(c, m.action)
		}
	}

	return c
}

func Handshake(code uint) []string {
	binary := strconv.FormatUint(uint64(code), 2)

	m := []modifier{
		{1, wink, nil},
		{2, doubleBlink, nil},
		{3, closeYourEyes, nil},
		{4, jump, nil},
		{5, "", func(binary string, c []string) []string {
			slices.Reverse(c)
			return c
		}},
	}

	var c []string
	for i := 0; i < len(m); i++ {
		c = m[i].apply(binary, c)
	}

	return []string(c)
}
